toss.py

```diff
@@ -42,59 +44,98 @@ def load_udp(args, initial_condition, lower_bounds, upper_bounds):
+def run_optimization(args, initial_conditions, lower_bounds, upper_bounds):
+
+    # Setup BFE machinery for paralellization (activates engines)
     multi_process_bfe = pg.mp_bfe() #pg.ipyparallel_bfe() #n=args.optimization.number_of_threads
     multi_process_bfe.resize_pool(args.optimization.number_of_threads)
     bfe = pg.bfe(multi_process_bfe)
 
-    # Setup population
-    pop = pg.population(prob=prob, size=args.optimization.population_size, b=bfe)
+    # Setup arrays for storing results:
+    champion_f_array = np.empty(args.problem.number_of_spacecrafts, dtype=np.float64)
+    champion_x_array = np.empty(args.problem.number_of_spacecrafts*len(lower_bounds), dtype=np.float64)
+    fitness_arr = np.empty((args.problem.number_of_spacecrafts, args.optimization.number_of_generations), dtype=np.float64)
 
-    # Evolve archipelago (Optimization process)
-    uda.set_bfe(bfe)
-    algo = pg.algorithm(uda)
-
-    algo.set_verbosity(1)
-    pop = algo.evolve(pop)
+    # Initiate timer of the optimization process
+    timer_start = time.time()
     
-    # Store champion fitness of each generation:
-    optimization_info = algo.extract(pg.gaco).get_log()
-    fitness_list = np.empty(args.optimization.number_of_generations,dtype=np.float64)
-    for generation, info in enumerate(optimization_info):
-        fitness_list[generation] = info[2]
-
-    # Logs for output
-    run_time_end = time.time()
-    run_time = run_time_end - run_time_start
-    champion_f = pop.champion_f
-    champion_x = pop.champion_x
-    print("Optimization run time: ", run_time)
-    print("Champion fitness value: ", champion_f) 
-    print("Champion chromosome: ", champion_x) 
+    # Optimize each spacecraft trajectory
+    for spacecraft_i in range(0, args.problem.number_of_spacecrafts):
+
+        # Setup udp
+        prob = load_udp(args, initial_conditions[spacecraft_i], lower_bounds, upper_bounds)
+
+        # Setup population
+        pop = pg.population(prob=prob, size=args.optimization.population_size, b=bfe)
+
+        # Setup uda
+        algo = load_uda(args, bfe)
+
+        # Evolve population
+        algo.set_verbosity(1)
+        pop = algo.evolve(pop)
+
+        # Store champion fitness of each generation:
+        optimization_info = algo.extract(pg.gaco).get_log()
+        fitness_list = np.empty(args.optimization.number_of_generations,dtype=np.float64)
+        for generation, info in enumerate(optimization_info):
+            fitness_list[generation] = info[2]
+
+        # Other logs for output
+        champion_f = pop.champion_f
+        champion_x = pop.champion_x
+
+        # Update boolean tensor (using trajectory resulting from champion chromosome)
+        if len(initial_conditions[spacecraft_i]) > 0:
+            spacecraft_info = np.hstack((initial_conditions[spacecraft_i], champion_x))
+        else:
+            spacecraft_info = champion_x
+
+        _, list_of_ode_objects, _ = compute_trajectory(spacecraft_info, args, compute_motion)
```

Comment:
Actually this one was a bit tricky. Since I couldn't figure out how to reach and manipulate the attributes of the udp class through the pygmo problem object, I had to move the updating of the tensor to outside the optimization problem. Because of this, we have to re-initialize the udp and corresponding population for each new spacecraft as we will have a new tensor to account for. Since the integration objects are managed in the udp as well, and since pygmo only returns the fitness and some statistics on the optimization process, this means that we have to re-compute the trajectory using the champion outside the optimization in order to update the tensor. Luckily we only need to do this once per spacecraft so it shouldn't take too much time (although re-initializing the population is probably not very optimal with respect to time).